src/v/archival/segment_reupload.cc

```diff
@@ -0,0 +1,251 @@
+            _end_inclusive = current_segment_end;
+        } else {
+            _end_inclusive = it->first.base_offset - model::offset{1};
+        }
+    }
+}
+
+segment_collector::lookup_result
+segment_collector::find_next_compacted_segment(model::offset start_offset) {
+    const auto& segment_set = _log->segments();
+    auto it = segment_set.lower_bound(start_offset);
+    if (
+      it == segment_set.end() && start_offset < _log->offsets().start_offset) {
+        it = segment_set.begin();
+    }
+
+    if (it == segment_set.end()) {
+        vlog(
+          archival_log.debug,
+          "Finding next compacted segment for {}: can't find segment after "
+          "offset: {}",
+          _manifest->get_ntp(),
+          start_offset);
+        return {};
+    }
+
+    const auto& segment = *it;
+    if (segment->finished_self_compaction()) {
+        vlog(
+          archival_log.trace,
+          "Found compacted segment for ntp {}: {}",
+          _manifest->get_ntp(),
+          segment);
+        return {.segment = segment, .ntp_conf = &_log->config()};
+    }
+    vlog(
+      archival_log.debug,
+      "Finding next compacted segment for {}: no compacted "
+      "segments after offset: {}",
+      _manifest->get_ntp(),
+      start_offset);
+    return {};
+}
+
+model::offset segment_collector::begin_inclusive() const {
+    return _begin_inclusive;
+}
+
+model::offset segment_collector::end_inclusive() const {
+    return _end_inclusive;
+}
+
+const storage::ntp_config* segment_collector::ntp_cfg() const {
+    return _ntp_cfg;
+}
+
+bool segment_collector::can_replace_manifest_segment() const {
+    return _can_replace_manifest_segment && _begin_inclusive < _end_inclusive;
+}
+
+cloud_storage::segment_name segment_collector::adjust_segment_name() const {
+    vassert(
+      !_segments.empty(), "Cannot calculate segment name with no segments");
+
+    auto first = _segments.front();
+    auto file_name = first->filename();
+    auto meta = storage::segment_path::parse_segment_filename(file_name);
+    auto version = meta ? meta->version : storage::record_version_type::v1;
+
+    cloud_storage::segment_name name{};
+    if (_begin_inclusive == first->offsets().base_offset) {
+        auto orig_path = std::filesystem::path(file_name);
+        name = cloud_storage::segment_name(orig_path.filename().string());
+        vlog(archival_log.debug, "Using original segment name: {}", name);
+    } else {
+        auto path = storage::segment_path::make_segment_path(
+          *_ntp_cfg, _begin_inclusive, first->offsets().term, version);
+        name = cloud_storage::segment_name(path.filename().string());
+        vlog(archival_log.debug, "Using adjusted segment name: {}", name);
+    }
+
+    return name;
+}
+
+void segment_collector::align_begin_offset_to_manifest() {
+    if (_begin_inclusive >= _manifest->get_last_offset()) {
+        return;
+    }
+
+    if (_begin_inclusive < _manifest->get_start_offset().value()) {
```

Comment:
can this optional be empty or is it an invariant that it will hold a value when it's used here?

Reply:
Yes, we have a check at https://github.com/redpanda-data/redpanda/pull/6565/files#diff-2bf47491353666e0f8335ea7eb62c6c2ad8ee21e48c0b29b1495770ddc10d087R30 before calling `align...` that the manifest is not empty, if it is non empty then `get_start_offset()` will always hold a value.